checkers/typeDefFirst_checker.go

```diff
@@ -0,0 +1,75 @@
+package checkers
+
+import (
+	"go/ast"
+	"go/token"
+	"strings"
+
+	"github.com/go-critic/go-critic/checkers/internal/astwalk"
+	"github.com/go-critic/go-critic/framework/linter"
+)
+
+func init() {
+	var info linter.CheckerInfo
+	info.Name = "typeDefFirst"
+	info.Tags = []string{"style", "experimental"}
+	info.Summary = "File-scoped checker, that requires type definition before its method definitions"
+	info.Before = `
+func (r rec) Method() {}
+type rec struct{}
+`
+	info.After = `
+type rec struct{}
+func (r rec) Method() {}
+`
+	collection.AddChecker(&info, func(ctx *linter.CheckerContext) linter.FileWalker {
+		return &typeDefFirstChecker{
+			ctx: ctx,
+		}
+	})
+}
+
+type typeDefFirstChecker struct {
+	astwalk.WalkHandler
+	ctx *linter.CheckerContext
+}
+
+func (t *typeDefFirstChecker) WalkFile(f *ast.File) {
```

Comment:
We have a convention to have a `c` for every checker receiver. This makes checker renaming easier (and we do rename experimental checkers from time to time before they get stable). It also makes the code more consistent.